Bugfix/issue352

slimboyfat opened the pull request:
This PR addresses two issues:

- Access to shared field reactor.kafka.receiver.internals.DefaultKafkaReceiver#consumerHandler w/o synchronization. 
  
- Reference to new consumerHandler might be overwritten with null by asynchronous cleanup action initiated by termination of previous subscription.
During a cleanup action consumerHandler reference is set to null only if reference wasn't changed (there is no other ConsumerHandler was created in the middle) 

Details.

Reference to consumerHandler is assigned from subscriber thread context, finalized from reactor.kafka.receiver.internals.ConsumerHandler#eventScheduler and might be accessed from any other thread via reactor.kafka.receiver.KafkaReceiver#doOnConsumer.
  
During a cleanup action consumerHandler reference is set to null only if reference wasn't changed (there is no other ConsumerHandler was created in the middle).

Wrapping consumerHandler into AtomicReference gives both: thread-safety and compareAndSwap for reference change.

pivotal-cla commented:
@slimboyfat Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-kafka&pullRequestId=355)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=reactor/reactor-kafka&pullRequestId=355) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@slimboyfat Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-kafka&pullRequestId=355)!

garyrussell commented:
Thanks for the contribution.

Copyright needs updating

>Run './gradlew :spotlessApply' to fix these violations.

Use `./gradlew clean check` to build locally.

slimboyfat commented:
> Copyright needs updating

I fixed copyright in [b9b868b](https://github.com/reactor/reactor-kafka/pull/355/commits/b9b868b52a2f58bc6db32ff95e9d8c48ab6e218d)

Also, I fixed some tests in [5081feb](https://github.com/reactor/reactor-kafka/pull/355/commits/5081feb6c002023d902006fea22edb16b0778a1e)
